diff --git a/CMakeLists.txt b/CMakeLists.txt
index 314c72fe8..f102b7710 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -597,14 +597,14 @@ add_subdirectory(unsupported)
 add_subdirectory(demos EXCLUDE_FROM_ALL)
 
 # must be after test and unsupported, for configuring buildtests.in
-add_subdirectory(scripts EXCLUDE_FROM_ALL)
+#add_subdirectory(scripts EXCLUDE_FROM_ALL)
 
 # TODO: consider also replacing EIGEN_BUILD_BTL by a custom target "make btl"?
 if(EIGEN_BUILD_BTL)
   add_subdirectory(bench/btl EXCLUDE_FROM_ALL)
 endif()
 
-find_package(CLANG_FORMAT 9 EXACT)
+#find_package(CLANG_FORMAT 9 EXACT)
 if(CLANG_FORMAT_FOUND)
 set(FORMAT_SOURCES)
 list(APPEND FORMAT_SUBDIRS blas bench demos "doc" Eigen include lapack scripts share unsupported test failtest)
@@ -631,9 +631,9 @@ if(EIGEN_BUILD_TESTING)
   ei_testing_print_summary()
 endif()
 
-message(STATUS "")
-message(STATUS "Configured Eigen ${EIGEN_VERSION_NUMBER}")
-message(STATUS "")
+#message(STATUS "")
+#message(STATUS "Configured Eigen ${EIGEN_VERSION_NUMBER}")
+#message(STATUS "")
 
 if(PROJECT_IS_TOP_LEVEL)
   string(TOLOWER "${CMAKE_GENERATOR}" cmake_generator_tolower)
