diff --git a/CMake/Version.cmake b/CMake/Version.cmake
index 4825d18b..aa4e8989 100644
--- a/CMake/Version.cmake
+++ b/CMake/Version.cmake
@@ -3,26 +3,26 @@ set(VERSION_MINOR 0)
 set(VERSION_PATCH 6)
 set(VERSION_COMMIT 0)
 
-find_program(GIT git)
-if(GIT AND EXISTS "${CMAKE_CURRENT_SOURCE_DIR}/.git")
-  execute_process(
-      COMMAND ${GIT} describe --tags
-      WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
-      OUTPUT_VARIABLE GIT_DESCRIBE_DIRTY
-      OUTPUT_STRIP_TRAILING_WHITESPACE
-      RESULT_VARIABLE GIT_DESCRIBE_RESULT
-  )
-
-  if(GIT_DESCRIBE_RESULT EQUAL 0)
-    string(REGEX REPLACE "^v([0-9]+)\\..*" "\\1" VERSION_MAJOR "${GIT_DESCRIBE_DIRTY}")
-    string(REGEX REPLACE "^v[0-9]+\\.([0-9]+).*" "\\1" VERSION_MINOR "${GIT_DESCRIBE_DIRTY}")
-    string(REGEX REPLACE "^v[0-9]+\\.[0-9]+\\.([0-9]+).*" "\\1" VERSION_PATCH "${GIT_DESCRIBE_DIRTY}")
-    string(REGEX REPLACE "^v[0-9]+\\.[0-9]+\\.[0-9]+\\-([0-9]+).*" "\\1" VERSION_COMMIT "${GIT_DESCRIBE_DIRTY}")
-  else()
-    message(WARNING "git describe failed with exit code: ${GIT_DESCRIBE_RESULT}")
-  endif()
-else()
-  message(WARNING "git is not found")
-endif()
-
-message(STATUS "Proceeding with version: ${VERSION_MAJOR}.${VERSION_MINOR}.${VERSION_PATCH}.${VERSION_COMMIT}")
+#find_program(GIT git)
+#if(GIT AND EXISTS "${CMAKE_CURRENT_SOURCE_DIR}/.git")
+#  execute_process(
+#      COMMAND ${GIT} describe --tags
+#      WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
+#      OUTPUT_VARIABLE GIT_DESCRIBE_DIRTY
+#      OUTPUT_STRIP_TRAILING_WHITESPACE
+#      RESULT_VARIABLE GIT_DESCRIBE_RESULT
+#  )
+#
+#  if(GIT_DESCRIBE_RESULT EQUAL 0)
+#    string(REGEX REPLACE "^v([0-9]+)\\..*" "\\1" VERSION_MAJOR "${GIT_DESCRIBE_DIRTY}")
+#    string(REGEX REPLACE "^v[0-9]+\\.([0-9]+).*" "\\1" VERSION_MINOR "${GIT_DESCRIBE_DIRTY}")
+#    string(REGEX REPLACE "^v[0-9]+\\.[0-9]+\\.([0-9]+).*" "\\1" VERSION_PATCH "${GIT_DESCRIBE_DIRTY}")
+#    string(REGEX REPLACE "^v[0-9]+\\.[0-9]+\\.[0-9]+\\-([0-9]+).*" "\\1" VERSION_COMMIT "${GIT_DESCRIBE_DIRTY}")
+#  else()
+#    message(WARNING "git describe failed with exit code: ${GIT_DESCRIBE_RESULT}")
+#  endif()
+#else()
+#  message(WARNING "git is not found")
+#endif()
+#
+#message(STATUS "Proceeding with version: ${VERSION_MAJOR}.${VERSION_MINOR}.${VERSION_PATCH}.${VERSION_COMMIT}")
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 21e90151..6969d4aa 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1,5 +1,5 @@
 # This is the legacy minimum version flatbuffers supported for a while.
-cmake_minimum_required(VERSION 2.8.12)
+cmake_minimum_required(VERSION 3.5)
 
 # CMake version 3.16 is the 'de-facto' minimum version for flatbuffers. If the
 # current cmake is older than this, warn the user and include the legacy file to
@@ -282,7 +282,7 @@ elseif(CMAKE_COMPILER_IS_GNUCXX)
       "${CMAKE_CXX_FLAGS} -std=c++0x")
   endif(CYGWIN)
   set(CMAKE_CXX_FLAGS
-    "${CMAKE_CXX_FLAGS} -Wall -pedantic -Werror -Wextra -Werror=shadow")
+    "${CMAKE_CXX_FLAGS} -Wall -pedantic -Werror -Wextra -Werror=shadow -Wno-stringop-overflow")
   set(FLATBUFFERS_PRIVATE_CXX_FLAGS "-Wold-style-cast")
   if (CMAKE_CXX_COMPILER_VERSION VERSION_GREATER 4.4)
     if (CMAKE_CXX_COMPILER_VERSION VERSION_GREATER 7.0)
