diff --git a/CMakeLists.txt b/CMakeLists.txt
index 5c25e757a..9896709ce 100755
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -7496,7 +7496,8 @@ TARGET_LINK_LIBRARIES(XNNPACK PRIVATE
   normalization
   operators
   microparams_init
-  subgraph)
+  subgraph
+  post-operation)
 TARGET_LINK_LIBRARIES(all_microkernels PRIVATE fp16)
 TARGET_LINK_LIBRARIES(packing PRIVATE fp16)
 TARGET_LINK_LIBRARIES(microparams_init PRIVATE fp16)
